server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java

```diff
@@ -0,0 +1,124 @@
+        private String index;
+        private ConstraintMode invocationMode;
+
+        ConstraintParams(BalancedShardsAllocator.Balancer balancer, BalancedShardsAllocator.ModelNode node,
+                        String index, ConstraintMode invocationMode) {
+            this.balancer = balancer;
+            this.node = node;
+            this.index = index;
+            this.invocationMode = invocationMode;
+        }
+    }
+
+    /**
+     * Evaluates configured allocation constraint predicates for given node - index combination; and returns
+     * a weight value based on the number of breached constraints.
+     *
+     * Constraint weight should be added to the weight calculated via weight function, to reduce priority
+     * of allocating on nodes with breached constraints.
+     */
+    public long weight(BalancedShardsAllocator.Balancer balancer, BalancedShardsAllocator.ModelNode node,
+                       String index, ConstraintMode invocationMode) {
+        int constraintsBreached = 0;
+        ConstraintParams params = new ConstraintParams(balancer, node, index, invocationMode);
+        for (Predicate<ConstraintParams> predicate: constraintPredicates) {
+            if (predicate.test(params)) {
+                constraintsBreached++;
+            }
+        }
+        return constraintsBreached * CONSTRAINT_WEIGHT;
+    }
+
+    /**
+     * Constraint to control number of shards of an index allocated on a single node.
+     *
+     * In current weight function implementation, when a node has significantly fewer shards than other nodes (e.g. during
+     * single new node addition or node replacement), its weight is much less than other nodes. All shard allocations
+     * at this time tend to land on the new node with skewed weight. This breaks index level balance in the cluster,
+     * by creating all shards of the same index on one node, often resulting in a hotspot on that node.
+     *
+     * This constraint is breached when balancer attempts to allocate more than average shards per index per node.
+     */
+    private Predicate<ConstraintParams> isIndexShardsPerNodeBreached() {
+        return (params) -> {
+            int currIndexShardsOnNode = params.node.numShards(params.index);
+            int allowedIndexShardsPerNode = (int) Math.ceil(params.balancer.avgShardsPerNode(params.index));
+            return (params.invocationMode != ConstraintMode.NONE &&
+                indexShardsPerNodeConstraint == params.invocationMode &&
+                currIndexShardsOnNode >= allowedIndexShardsPerNode);
+        };
+    }
+
+    /**
+     * {@link ConstraintMode} values define possible modes in which allocation constraints can be enabled on
+     * the balancer via cluster settings.
+     *   1. {@link ConstraintMode#NONE}: "none" disables allocation constraint for all decisions.
+     *   2. {@link ConstraintMode#UNASSIGNED}: "unassigned" enables constraints only for decisions to allocate unassigned shards.
```

Comment:
What about other balancer decisions like move and rebalance? Why not add those modes as well?

Reply:
Makes sense.